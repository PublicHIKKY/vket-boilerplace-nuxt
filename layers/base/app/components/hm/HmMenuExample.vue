<template>
  <section class="hm-menu-example">
    <Menu
      as="div"
      class="menu-container"
    >
      <MenuButton class="button">
        Menu
      </MenuButton>
      <MenuItems class="menu-items">
        <MenuItem
          v-slot="{ active }"
          class="item"
        >
          <span :class="`span ${active ? '-active' : ''}`">Hello</span>
        </MenuItem>
        <MenuItem
          v-slot="{ active }"
          class="item"
        >
          <span :class="`span ${active ? '-active' : ''}`">Howdy!</span>
        </MenuItem>
        <MenuItem
          v-slot="{ active }"
          class="item"
        >
          <span :class="`span ${active ? '-active' : ''}`">Yo!</span>
        </MenuItem>
        <MenuItem
          disabled
          class="item -disable"
        >
          <span>:D</span>
        </MenuItem>
      </MenuItems>
    </Menu>
  </section>
</template>

<script setup lang="ts">
import { Menu, MenuButton, MenuItems, MenuItem } from '@headlessui/vue'
</script>

<style lang="scss" scoped>
.menu-container {
  > .button {
    border: solid 1px;
    border-radius: 8px;
    padding: 4px;
  }

  > .menu-items {
    display: flex;
    flex-direction: column;
    padding: 8px 4px 8px 16px;
  }
}

.menu-items {
  > .item {
    border: solid;
    border-color: #0ff;
    padding: 4px;

    &.-active {
      border-color: #00f;
    }

    &.-disable {
      border: none;
    }
  }
}
</style>
