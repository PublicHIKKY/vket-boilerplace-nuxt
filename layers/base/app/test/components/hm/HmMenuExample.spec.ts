import { mount } from '@vue/test-utils'
import HmMenuExample from '#base/app/components/hm/HmMenuExample.vue'

test('ref component', () => {
  expect(HmMenuExample).toBeTruthy()
})

test('mount component', () => {
  const wrapper = mount(HmMenuExample)
  expect(wrapper.getCurrentComponent()).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()
})

describe('rendering test', () => {
  it(':menu', () => {
    const wrapper = mount(HmMenuExample)
    expect(wrapper.find('div[class="menu-container"]').exists()).toBe(true)
  })
  it(':button (open close button)', () => {
    const wrapper = mount(HmMenuExample)
    expect(wrapper.find('button[class="button"]').exists()).toBe(true)
    expect(wrapper.find('button[class="button"]').text()).toBe('Menu')
  })
  it(':menu-items', async () => {
    const wrapper = mount(HmMenuExample)

    // NOTE: クリックするとmenu-listが開かれ出現すること
    await wrapper.get('button[class="button"]').trigger('click')
    expect(wrapper.find('div[class="menu-items"]').exists()).toBe(true)

    // NOTE: 開いたあとにクリックするとmenu-listが閉じて消えること
    await wrapper.get('button[class="button"]').trigger('click')
    expect(wrapper.find('div[class="menu-items"]').exists()).toBe(false)
  })
  it(':item', async () => {
    const wrapper = mount(HmMenuExample)

    // NOTE: クリックするとmenu-listが開かれ出現すること
    await wrapper.get('button[class="button"]').trigger('click')
    expect(wrapper.find('span[class="span item"]').exists()).toBe(true)

    expect(wrapper.find('span[class="span item"]:nth-child(1)').text()).toBe(
      'Hello',
    )
    // NOTE: 上記は下記でもいい（at(n)）
    expect(wrapper.findAll('span[class="span item"]').at(0)?.text()).toBe(
      'Hello',
    )
    // NOTE: 上記は下記でもいい2（配列のn番目インデックス）
    expect(wrapper.findAll('span[class="span item"]')[0]?.text()).toBe('Hello')
    expect(wrapper.find('span[class="span item"]:nth-child(2)').text()).toBe(
      'Howdy!',
    )

    expect(wrapper.find('span[class="span item"]:nth-child(3)').text()).toBe(
      'Yo!',
    )
    expect(
      wrapper.find('span[class="item -disable"]:nth-child(4)').text(),
    ).toBe(':D')

    // NOTE: 開いたあとに子要素をクリックするとmenu-listが閉じて子要素も消えること
    await wrapper.get('button[class="button"]').trigger('click')
    expect(wrapper.find('div[class="menu-items"]').exists()).toBe(false)
    expect(wrapper.find('span[class="span item"]').exists()).toBe(false)
  })
})
