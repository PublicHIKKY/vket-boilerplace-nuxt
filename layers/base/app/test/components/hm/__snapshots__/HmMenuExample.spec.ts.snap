// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mount component 1`] = `
"<section data-v-76fe05f2="" class="hm-menu-example">
  <div data-v-76fe05f2="" data-headlessui-state="" class="menu-container"><button data-v-76fe05f2="" id="headlessui-menu-button-v-0" type="button" aria-haspopup="menu" aria-expanded="false" data-headlessui-state="" class="button"> Menu </button>
    <!---->
  </div>
</section>"
`;
